src/celestia/textinput.cpp

```diff
@@ -0,0 +1,169 @@
+#include <celutil/utf8.h>
+
+namespace celestia
+{
+
+std::string_view
+TextInput::getTypedText() const
+{
+    return m_text;
+}
+
+
+util::array_view<std::string>
+TextInput::getCompletion() const
+{
+    return m_completion;
+}
+
+
+int
+TextInput::getCompletionIndex() const
+{
+    return m_completionIdx;
+}
+
+
+CharEnteredResult
+TextInput::charEntered(const Simulation* sim, std::string_view c_p, bool withLocations)
+{
+    char c = c_p.front();
+    if (std::int32_t uc = 0;
+        UTF8Decode(c_p, uc) && !(uc <= WCHAR_MAX && std::iswcntrl(static_cast<std::wint_t>(uc))))
+    {
+        appendText(sim, c_p, withLocations);
+    }
+    else if (c == '\b')
+    {
+        m_completionIdx = -1;
+        if (!m_text.empty())
+        {
+#ifdef AUTO_COMPLETION
+            do
+            {
+#endif
+                for (;;)
+                {
+                    auto ch = static_cast<std::byte>(m_text.back());
+                    m_text.pop_back();
+                    // If the string is empty, or the removed character was
+                    // not a UTF-8 continuation byte 0b10xx_xxxx then we're
+                    // done.
+                    if (m_text.empty() || (ch & std::byte(0xc0)) != std::byte(0x80))
+                        break;
+                }
+
+                m_completion.clear();
+                if (!m_text.empty())
+                        sim->getObjectCompletion(m_completion, m_text, withLocations);
+#ifdef AUTO_COMPLETION
+            } while (!typedText.empty() && typedTextCompletion.size() == 1);
+#endif
+        }
+    }
+    else if (c == '\011') // TAB
+    {
+        if (m_completionIdx + 1 < (int) m_completion.size())
+            m_completionIdx++;
+        else if ((int) m_completion.size() > 0 && m_completionIdx + 1 == (int) m_completion.size())
+            m_completionIdx = 0;
+        if (m_completionIdx >= 0) {
+            std::string::size_type pos = m_text.rfind('/', m_text.length());
+            if (pos != std::string::npos)
+            {
+                m_text.resize(pos + 1);
+                m_text.append(m_completion[m_completionIdx]);
+            }
+            else
+            {
+                m_text = m_completion[m_completionIdx];
+            }
+        }
+    }
+    else if (c == '\177') // BackTab
+    {
+        if (m_completionIdx > 0)
+            m_completionIdx--;
+        else if (m_completionIdx == 0)
+            m_completionIdx = m_completion.size() - 1;
+        else if (m_completion.size() > 0)
+            m_completionIdx = m_completion.size() - 1;
+        if (m_completionIdx >= 0) {
+            std::string::size_type pos = m_text.rfind('/', m_text.length());
+            if (pos != std::string::npos)
+            {
+                m_text.resize(pos + 1);
+                m_text.append(m_completion[m_completionIdx]);
+            }
+            else
+            {
+                m_text = m_completion[m_completionIdx];
+            }
```

Comment:
```suggestion
            if (pos != std::string::npos)
                m_text.resize(pos + 1);
            m_text = m_completion[m_completionIdx];
```

Reply:
That doesn't do the same thing... think it would be:

```cpp
m_text.resize(pos == std::string::npos ? 0 : pos + 1);
m_text.append(m_completion[m_completionIdx]);
```